src/graph/graph.c

```diff
@@ -170,6 +237,146 @@ Node* Graph_GetNode(const Graph *g, int id) {
     return &block->nodes[id%NODEBLOCK_CAP];
 }
 
+void Graph_DeleteNodes(Graph *g, int *IDs, size_t IDCount) {
+    assert(g && IDs);
+    if(IDCount == 0) return;
+
+    typedef struct {
+        int nodeID;         // Node being deleted.
+        int replacementID;  // Node taking over.
+        bool delete;        // No need to replace, simply delete.
+    } Replacement;
+
+    Replacement *replacements = malloc(sizeof(Replacement) * IDCount);
+
+    /* Allocate replacement candidates. */
+    for(int i = 0; i < IDCount; i++) {
```

Comment:
I feel like this approach is more elaborate than it needs to be. Can we just:

```
Sort the IDs array in descending order
int innermost = IDCount - 1;
for (i = 0; i < innermost; i ++) {
  for (j = IDs[i] - 1; j > IDs[i + 1]; j --) {
    replace node at IDs[innermost] with node at j, update matrices
    innermost --;
    if (innermost <= i) {
      max_remaining = j;
      break;
    }
  }
}

resize matrix to max_remaining
```
I am likely overlooking some edge cases and getting some offsets wrong, but generally I feel like if we sort the nodes to be deleted, we can quickly tell which higher-ID nodes are safe candidates to use as replacements.
Then we can swap them immediately, and continue doing so until there are no to-be-deleted IDs less than the updated node count.
At that point, we can resize the matrix down to the ID of the last replacement node we swapped in.

Reply:
At first I thought about sorting which will defiantly make this code clearer, once we have a decent sorting algorithm in place there are additional areas in the code which can benefit from it.
I'll keep this as is for now and return to it once we find an optimised sorting algorithm.